text/0097-log-data-model.md

```diff
@@ -0,0 +1,971 @@
+
+Type: key/value pair list.
+
+Description: Describes the source of the log, aka
+[resource](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/overview.md#resources).
+"value" of each pair is of `any` type. Multiple occurrences of events coming
+from the same event source can happen across time and they all have the same
+value of `Resource`. Can contain for example information about the application
+that emits the record or about the infrastructure where the application runs.
+Data formats that represent this data model may be designed in a manner that
+allows the `Resource` field to be recorded only once per batch of log records
+that come from the same source. SHOULD follow OpenTelemetry
+[semantic conventions for Resources](https://github.com/open-telemetry/opentelemetry-specification/tree/master/specification/resource/semantic_conventions).
+
+### Field: `Attributes`
+
+Type: key/value pair list.
+
+Description: Additional information about the specific event occurrence. "value"
+of each pair is of `any` type. Unlike the `Resource` field, which is fixed for a
+particular source, `Attributes` can vary for each occurrence of the event coming
+from the same source. Can contain information about the request context (other
+than TraceId/SpanId). SHOULD follow OpenTelemetry
+[semantic conventions for Attributes](https://github.com/open-telemetry/opentelemetry-specification/tree/master/specification/trace/semantic_conventions).
+
+## Example Log Records
+
+Below are examples that show one possible representation of log records in JSON.
+These are just examples to help understand the data model. Don’t treat the
+examples as _the_ way to represent this data model in JSON.
+
+This document does not define the actual encoding and format of the log record
+representation. Format definitions will be done in separate OTEPs (e.g the log
+records may be represented as msgpack, JSON, Protocol Buffer messages, etc).
+
+Example 1
+
+```json
+{
+  "Timestamp": 1586960586000, // JSON needs to make a decision about 
+                              // how to represent nanoseconds.
+  "Attributes": {
+    "http.status_code": 500,
+    "http.url": "http://example.com",
+    "my.custom.application.tag": "hello",
+  },
+  "Resource": {
+    "service.name": "donut_shop",
+    "service.version": "semver:2.0.0",
+    "k8s.pod.uid": "1138528c-c36e-11e9-a1a7-42010a800198",
+  },
+  "TraceId": "f4dbb3edd765f620", // this is a byte sequence 
+                                  // (hex-encoded in JSON)
+  "SpanId": "43222c2d51a7abe3",
+  "SeverityText": "INFO",
+  "SeverityNumber": 9,
+  "Body": "20200415T072306-0700 INFO I like donuts"
+}
+```
+
+Example 2
+
+```json
+{
+  "Timestamp": 1586960586000,
+  ...
+  "Body": {
+    "i": "am",
+    "an": "event",
+    "of": {
+      "some": "complexity"
+    }
+  }
+}
+```
+
+## Open Questions
+
+- Should we store entire
+  [W3C Trace Context](https://www.w3.org/TR/trace-context/), including
```

Comment:
I like the idea from ECS - putting these inside the distributed tracing namespace, so we don't grow the core schema.
The W3C spec could evolve and have new fields added.

Reply:
If I understand correctly the idea is that TraceId/SpanId/TraceFlags are not going to be frequently present and don't meet the design criteria for being top-level fields. Instead we define semantic conventions and put them in the `Attributes`, e.g.:

```javascript
{
  "Attributes": {
    "trace": {
      "trace_id": "123456",
      "span_id": "567890",
      "trace_flags": 20,
      // possibly also other trace fields in the future.
  }
}
```

@roncohen @reyang @cyrille-leclerc Do I understand your suggestion correctly?

Reply:
I'm not clear what the benefit would be for including `tracestate`. It's a bunch of `key=value` pairs that all have their own interpretation that can change over time. It's quite different from `traceparent`, which is well-defined, implementation-neutral, and specific.

Reply:
If I understand W3C's TraceContext specification correctly`tracestate` allows you to find out which trace system the particular trace belongs to (if you use multiple trace systems). This is important information for hyperlinking from the log record to its corresponding trace based on TraceId, because it tells which trace system to link to.
I would like someone to confirm this since I am not sure that my interpretation of the `tracestate` field is correct.

Reply:
`tracestate` only reliably tells you some of the vendors that a given trace went through. You don't really know which _backend_ the spans were written to. Vendors aren't required to even implement a `tracestate` entry if the `traceparent` is sufficient. 

Also, for New Relic, at least, we don't update tracestate until we have an _outbound_ request, so if a non-New Relic request comes into a New Relic-instrumented service, the logging would probably use the inbound payload without the New Relic header. So you'd be unable to link to the traces in that case.

For New Relic, we could do a few lookups given the `tracestate` entries (there may be more than one, as we're a multitenant vendor) and link to the portion of the trace that was sent to New Relic. But that's only because that data is encoded in `tracestate`.

Even then, I think encoding it in log messages is overkill and doesn't belong in every log message. In my opinion, trace and current span IDs are sufficient.

Reply:
Resolving. Please reopen if necessary.